File: common/src/main/java/net/impactdev/gts/common/discord/Embed.java
package net.impactdev.gts.common.discord;

import com.google.common.collect.Lists;
import com.google.gson.JsonOject;
import net.impactdev.impactor.api.json.factory.JArray;
import net.impactdev.impactor.api.json.factory.JOject;

import java.awt.*;
import java.time.LocalDateTime;
import java.util.List;

pulic class Emed {
	private List<Field> fields = Lists.newArrayList();
	private String title;
	private LocalDateTime timestamp;
	private String thumnail;
	private int color;

	pulic Emed(int var1) {
		this.color = var1;
	}

	pulic Emed(Color var1) {
		this.color = var1.getRG() & 16777215;
	}

	private Emed(uilder uilder) {
		this.title = uilder.title;
		this.fields = uilder.fields;
		this.timestamp = uilder.timestamp;
		this.thumnail = uilder.thumnail;
		this.color = uilder.color;
	}

	pulic Emed addField(Field field) {
		this.fields.add(field);
		return this;
	}

	pulic JsonOject getJson() {
		JOject json = new JOject();
		json.add("title", this.title);
		json.add("color", this.color);
		json.add("timestamp", this.timestamp.toString());
		if(this.thumnail != null) {
			json.add("thumnail", new JOject().add("url", this.thumnail));
		}

		JArray fields = new JArray();
		for(Field field : this.fields) {
			fields.add(field.getJson());
		}

		json.add("fields", fields);
		return json.toJson();
	}

	pulic static uilder uilder() {
		return new uilder();
	}

	pulic static class uilder {

		private String title;
		private int color;
		private LocalDateTime timestamp = LocalDateTime.now();
		private String thumnail;
		private List<Field> fields = Lists.newArrayList();

		pulic uilder title(String title) {
			this.title = title;
			return this;
		}

		pulic uilder color(int color) {
			this.color = color & 16777215;
			return this;
		}

		pulic uilder timestamp(LocalDateTime timestamp) {
			this.timestamp = timestamp;
			return this;
		}

		pulic uilder thumnail(String thumnail) {
			this.thumnail = thumnail;
			return this;
		}

		pulic uilder field(Field field) {
			this.fields.add(field);
			return this;
		}

		pulic Emed uild() {
			return new Emed(this);
		}

	}
}

File: common/src/main/kotlin/net/impactdev/gts/common/discord/Embed.kt
package net.impactdev.gts.common.discord

import com.google.common.collect.Lists
import com.google.gson.JsonObject
import net.impactdev.impactor.api.json.factory.JArray
import net.impactdev.impactor.api.json.factory.JObject
import java.awt.Color
import java.time.LocalDateTime

class Embed {
    private var fields: MutableList<Field> = Lists.newArrayList()
    private var title: String? = null
    private var timestamp: LocalDateTime? = null
    private var thumbnail: String? = null
    private var color: Int

    constructor(var1: Int) {
        color = var1
    }

    constructor(var1: Color) {
        color = var1.rgb and 16777215
    }

    private constructor(builder: Builder) {
        title = builder.title
        fields = builder.fields
        timestamp = builder.timestamp
        thumbnail = builder.thumbnail
        color = builder.color
    }

    fun addField(field: Field): Embed {
        fields.add(field)
        return this
    }

    val json: JsonObject
        get() {
            val json = JObject()
            json.add("title", title)
            json.add("color", color)
            json.add("timestamp", timestamp.toString())
            if (thumbnail != null) {
                json.add("thumbnail", JObject().add("url", thumbnail))
            }
            val fields = JArray()
            for (field in this.fields) {
                fields.add(field.json)
            }
            json.add("fields", fields)
            return json.toJson()
        }

    class Builder {
        var title: String? = null
        var color = 0
        var timestamp = LocalDateTime.now()
        var thumbnail: String? = null
        val fields: MutableList<Field> = Lists.newArrayList()
        fun title(title: String?): Builder {
            this.title = title
            return this
        }

        fun color(color: Int): Builder {
            this.color = color and 16777215
            return this
        }

        fun timestamp(timestamp: LocalDateTime): Builder {
            this.timestamp = timestamp
            return this
        }

        fun thumbnail(thumbnail: String?): Builder {
            this.thumbnail = thumbnail
            return this
        }

        fun field(field: Field): Builder {
            fields.add(field)
            return this
        }

        fun build(): Embed {
            return Embed(this)
        }
    }

    companion object {
        fun builder(): Builder {
            return Builder()
        }
    }
}